lib/ua_update_messaging.py

```diff
@@ -0,0 +1,317 @@
+    cfg: config.UAConfig
+) -> "Tuple[ContractExpiryStatus, int]":
+    """Return a tuple [ContractExpiryStatus, num_days]"""
+    if not cfg.is_attached:
+        return ContractExpiryStatus.NONE, 0
+
+    grace_period = defaults.CONTRACT_EXPIRY_GRACE_PERIOD_DAYS
+    pending_expiry = defaults.CONTRACT_EXPIRY_PENDING_DAYS
+    remaining_days = cfg.contract_remaining_days
+    if 0 <= remaining_days <= pending_expiry:
+        return ContractExpiryStatus.ACTIVE_EXPIRED_SOON, remaining_days
+    elif -grace_period <= remaining_days < 0:
+        return ContractExpiryStatus.EXPIRED_GRACE_PERIOD, remaining_days
+    elif remaining_days < -grace_period:
+        return ContractExpiryStatus.EXPIRED, remaining_days
+    return ContractExpiryStatus.ACTIVE, remaining_days
+
+
+def _write_template_or_remove(msg: str, tmpl_file: str):
+    """Write a template to tmpl_file.
+
+    When msg is empty, remove both tmpl_file and the generated msg.
+    """
+    if msg:
+        util.write_file(tmpl_file, "\n" + msg)
+    else:
+        util.remove_file(tmpl_file)
+        if tmpl_file.endswith(".tmpl"):
+            util.remove_file(tmpl_file.replace(".tmpl", ""))
+
+
+def _write_esm_service_msg_templates(
+    cfg: config.UAConfig,
+    ent: entitlements.base.UAEntitlement,
+    expiry_status: ContractExpiryStatus,
+    remaining_days: int,
+    pkgs_file: str,
+    no_pkgs_file: str,
+    motd_pkgs_file: str,
+    motd_no_pkgs_file: str,
+):
+
+    pkgs_msg = no_pkgs_msg = motd_pkgs_msg = motd_no_pkgs_msg = ""
+    tmpl_prefix = ent.name.upper().replace("-", "_")
+    tmpl_pkg_count_var = "{{{}_PKG_COUNT}}".format(tmpl_prefix)
+    tmpl_pkg_names_var = "{{{}_PACKAGES}}".format(tmpl_prefix)
+    if ent.application_status()[0] == ApplicationStatus.ENABLED:
+        if expiry_status == ContractExpiryStatus.ACTIVE_EXPIRED_SOON:
+            pkgs_msg = MESSAGE_CONTRACT_EXPIRED_SOON_TMPL.format(
+                title=ent.title,
+                remaining_days=remaining_days,
+                url=defaults.BASE_UA_URL,
+            )
+            # Same cautionary message when contract is about to expire
+            motd_pkgs_msg = motd_no_pkgs_msg = no_pkgs_msg = pkgs_msg
+        elif expiry_status == ContractExpiryStatus.EXPIRED_GRACE_PERIOD:
+            grace_period_remaining = (
+                defaults.CONTRACT_EXPIRY_GRACE_PERIOD_DAYS + remaining_days
+            )
+            pkgs_msg = MESSAGE_CONTRACT_EXPIRED_GRACE_PERIOD_TMPL.format(
+                expired_date=cfg.contract_expiry_datetime.strftime("%d %b %Y"),
+                remaining_days=grace_period_remaining,
+                url=defaults.BASE_UA_URL,
+            )
+            # Same cautionary message when in grace period
+            motd_pkgs_msg = motd_no_pkgs_msg = no_pkgs_msg = pkgs_msg
+        elif expiry_status == ContractExpiryStatus.EXPIRED:
+            series = util.get_platform_info()["series"]
+            if ent.name == "esm-apps" and series in ("trusty", "xenial"):
```

Comment:
Why only for "esm-apps" here?

Reply:
Smart man, I think I actually need to separate this to a separate msg file because it needs to be the footer after both APPS and Infra expiry msgs.